### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mhutter**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

### CODE REVIEW

File: class/defaults.yml [hunk 1]
```diff
@@ -1,4 +1,38 @@
 parameters:
   machineset_egress_cidr_operator:
     =_metadata: {}
+
     namespace: syn-machineset-egress-cidr-operator
+
+    images:
+      openshift:
```
Comment (ccremer) [inline class/defaults.yml:8]:
```suggestion
      operator:
```

File: docs/modules/ROOT/pages/references/parameters.adoc [hunk 2]
```diff
@@ -17,3 +17,35 @@ The namespace in which to deploy this component.
...
 ----
+
+== `images`
+
+[horizontal]
+type:: object
+default::
+[source,yaml]
+----
+openshift4:
```
Comment (ccremer) [inline docs/modules/ROOT/pages/references/parameters.adoc:28]:
```suggestion
operator:
```

File: component/main.jsonnet [hunk 3]
```diff
@@ -5,6 +5,61 @@ local inv = kap.inventory();
 // The hiera parameters for the component
 local params = inv.parameters.machineset_egress_cidr_operator;
 
+local namespace = kube.Namespace(params.namespace);
+
+local serviceAccount = kube.ServiceAccount('machineset-egress-cidr-operator') {
+  metadata+: {
+    namespace: params.namespace,
+  },
```
Comment (ccremer) [inline component/main.jsonnet:13]: You don't need to explicitly define the namespaces here. The ArgoCD App object will apply the namespace to all objects that haven't explicitly defined it.
In other words, you only need to define the ns if it's not the same as `params.namespace`

File: component/main.jsonnet [hunk 4]
```diff
@@ -5,6 +5,61 @@ local inv = kap.inventory();
...
+
+local clusterRoleBinding = kube.ClusterRoleBinding('machineset-egress-cidr-operators') {
+  subjects_: [serviceAccount],
+  roleRef_: clusterRole,
+};
+
+local roleBindnig = kube.RoleBinding('machineset-egress-cidr-operators') {
+  metadata+: {
+    namespace: params.namespace,
+  },
```
Comment (ccremer) [inline component/main.jsonnet:28]: same here

File: component/main.jsonnet [hunk 5]
```diff
@@ -5,6 +5,61 @@ local inv = kap.inventory();
...
+    name: 'leader-election-role',
+  },
+};
+
+local deployment = kube.Deployment('machineset-egress-cidr-operators') {
+  metadata+: {
+    namespace: params.namespace,
+    labels: {
+      'app.kubernetes.io/name': 'machineset-egress-cidr-operator',
+      'app.kubernetes.io/managed-by': 'syn',
```
Comment (ccremer) [inline component/main.jsonnet:41]: I recommend to define the common labels in the parameters and then you can reuse it for every object like `lables: params.labels,`

You also might want to add more labels (see https://syn.tools/syn/references/labels-guide.html)

**Comment by mhutter** [comment]
@ccremer now I ended up with
```
  metadata+: {
    labels+: params.labels,
  },
```
on all objects, is there a way to streamline this (except for implementing a func that adds them to all objects in a list)?

**Comment by mhutter** [comment]
@ccremer now I ended up with
```
  metadata+: {
    labels+: params.labels,
  },
```
on all objects, is there a way to streamline this (except for implementing a func that adds them to all objects in a list)?

File: component/main.jsonnet [hunk 6]
```diff
@@ -5,6 +5,67 @@ local inv = kap.inventory();
...
+  metadata+: {
+    labels+: params.labels,
+  },
+};
+
+local clusterRole = kube.ClusterRole('machineset-egress-cidr-operator') {
+  metadata+: {
+    labels+: params.labels,
+  },
+  rules: params.clusterRole_rules,
```
Comment (mhutter) [inline component/main.jsonnet:24]: Search & replace collateral damage
```suggestion
  rules: params.cluster_role_rules,
```

**Comment by mhutter** [comment]
Found something easier :-) Happy for now

**Pull request merged**